### PR TITLE
EA: require size instead of __len__

### PR DESCRIPTION
**Pull request by jbrockmendel**
Discussed [here](https://github.com/jbrockmendel/pandas/pull/3)

Change the EA interface to require `size` instead of `__len__`.

Doing this separately from #27142 makes it much simpler to explain to downstream authors what is changing for them.  After this, the rest of the changes necessary for (limited) 2D support should not affect downstream authors.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
looks like the docs thing is driven by a URLError

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -40,6 +40,7 @@ Other enhancements
 Backwards incompatible API changes
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+- The :class:`ExtensionArray` interface now requires the author to implement ``size`` instead of ``__len__``.  The ``size`` method must _not_ depend on either ``__len__`` or ``shape`` (:issue:`????`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: you can use this issue number

**Review by jreback — CHANGES_REQUESTED** [review]
do we need tests to guarantee this method exists in EA? e.g. for decmial / json / arrow bool does it?

**Comment by jbrockmendel** [comment]
> do we need tests to guarantee this method exists in EA? e.g. for decmial / json / arrow bool does it?

This PR updates all of our internal (including tests.extension) EAs appropriately, and this is indirectly tested.

I've been giving some thought to how to test the "size must not depend on shape or len" requirement.  Thinking of something along the lines of a test that monkeypatches `shape` and `__len__` to raise, then calls `size`.  That leaves a corner case where an author could write a `size` property that only calls `shape` or `__len__` in some cases but not others.  Not sure this is worth worrying about.  Thoughts?

**Comment by TomAugspurger** [comment]
Can we schedule a call sometime this week to talk through this?

**Comment by jorisvandenbossche** [comment]
> Can we schedule a call sometime this week to talk through this?

Yes that is fine for me.

I am -1 on this PR until there is a clear vision and agreement on the next steps.

**Comment by jbrockmendel** [comment]
Wednesday or Thursday after 9 AM Pacific Time work for me

**Comment by jorisvandenbossche** [comment]
Both work for me (although I would prefer one hour earlier or 1.5 hour later)

**Comment by jreback** [comment]
these times are ok

**Comment by jbrockmendel** [comment]
An hour earlier on Wednesday works for me.  Pending Tom's schedule, let's go for Wednesday at 15:00 UTC.

**Comment by TomAugspurger** [comment]
I should be OK tomorrow (Wednesday) 1 hour earlier (8:00 Pacific / 16:00
CET) or 2.5 hours later (11:30 Pacific / 19:30 CET).

I'll be busy from 12:30 - 1:30 Central.

Free most of Friday.

**Comment by jreback** [comment]
did we agree in a date / time?

**Comment by TomAugspurger** [comment]
I think 15:00 UTC / 11:00 Eastern / 8:00 Pacific.

**Comment by TomAugspurger** [comment]
cc @pandas-dev/pandas-core we're having a call on this in ~25 minutes.

https://meet.google.com/gvh-okpg-qux

https://docs.google.com/document/d/1k_E_1oSV9VNHgGzepdeCyFdju8ZaXDwI3CvDYkhOtQ8/edit?usp=sharing

**Comment by simonjayhawkins** [comment]
Sorry can't hear no sound headphones not working.

**Comment by jbrockmendel** [comment]
Closing to clear the queue.

**Pull request closed without merging**